### PR TITLE
[CR] Make all vehicle parts foldable *but* not all vehicles

### PR DESCRIPTION
**Pull request by BevapDin**
```SUMMARY: Feature "Make all vehicle parts foldable, but not all vehicles"```

Foldable in out case doesn't necessarily mean it's physically changed, it just means it can be converted into an item. And that in turn can be done for all vehicles.

The current system is inconsistent as it only checks whether all parts of the vehicle can be "folded":

- a "vehicle" consisting of a frame, casters and a box can not be converted (aka folded) into an item as the frame is not foldable.
- a vehicle consisting of 10x10 foldable frames can be converted into an item.
- a foldable vehicle is suddenly not foldable after installing things like a motorcycle kickstand.

This PR changes it so all vehicle parts are considered foldable. If they don't have a foldable volume defined, the game will use the volume of the part item itself.

It also adds a limit to the size (500 L, 500 kg) of the folded vehicle. Larger vehicles can not be folded at all. Smaller vehicles can always be folded. It also adds a check for whether the resulting item was successfully added to the map (which may fail if there is no space for the item).

### CODE REVIEW

File: src/vehicle_use.cpp [hunk 1]
```diff
@@ -769,15 +776,55 @@ void vehicle::use_controls( const tripoint &pos )
...
+
+    units::mass weight = 0_gram;
+    units::volume volume = 0_ml;
+    for( const auto &part : parts ) {
+        volume += part.info().folded_volume ? *part.info().folded_volume : part.get_base().volume();
+        weight += part.get_base().weight();
+    }
+
+    // arbitrary limit
+    if( volume >= 500_liter ) {
```
Comment (AMurkin) [inline src/vehicle_use.cpp:805]: Can these limits respect user preferred units? In messages at least.

**Comment by kevingranade** [comment]
I'm not against the shift from per-part whitelisting to whole-vehicle evaluation in general.
I understand your point about this feature not representing folding, but if so at least the user-facing messaging, and ideally the internals as well, need to shift from using "folding" as the term to a more accurate term.  I don't have a slam-dunk term handy, but "packable" at least partially captures the intent of "can pick it up and stash it somewhere".

I'm not super happy with the criteria of 500L, I think a max weight is fairly reasonable, though much smaller than 500kg, possibly based on what the player can lift, but a sum of the volumes of the components doesn't tell us much about whether the vehicle is wieldy enough to meaningfully manipulate. Vehicle frames range from 8.75L to 20L, so no remotely sane number of them is going to meaningfully contribute to reaching the volume cap, but a 3m long motorcycle doesn't seem like a candidate for being packable.  There's a very obvious example of picking up a "small" 2x2 tile cart and then walking through a door with it.  That shouldn't work but is definitely allowed by a weight and volume threshold.

I think we need to back up a bit and define what the ultimate criteria for packable includes.  The original foldable criteria was based on "pick it up and put it in your backpack", but I'm guessing it's shifted a bit.  I can see extending it to "pick it up and throw it over your shoulder" i.e. a triathalon bike, but I'm not clear what the other use cases are, especially since dragging and hoisting vehicles is an option now.

**Comment by esotericist** [comment]
> but a sum of the volumes of the components doesn't tell us much about whether the vehicle is wieldy enough to meaningfully manipulate.

i think the crux of the problem you almost get at here is that we do not currently have any concept of how much volume the *vehicle* occupies, which is nearly always going to be *much* larger than the simple sum of the components. i don't know how we could derive that information from the infrastructure we currently have, but i think knowing that is a crucial requirement to evaluating if a vehicle can be carried without being manipulated to a different configuration (or even whether a different configuration could achieve carry-able)

File: src/vehicle_use.cpp [hunk 2]
```diff
@@ -796,24 +843,26 @@ bool vehicle::fold_up()
...
 
+    const ret_val<item> folded = get_folded();
+    if( !folded.success() ) {
+        add_msg( "%s", folded.str() );
+        return false;
+    }
+
+    // add_item_or_charges returns a reference to the placed item, or a null item if the
+    // item could not be placed anywhere (usually because it's too large).
+    if( g->m.add_item_or_charges( global_part_pos3( 0 ), folded.value() ).is_null() ) {
```
Comment (ZhilkinSerg) [inline src/vehicle_use.cpp:854]:
```suggestion
    if( get_map().add_item_or_charges( global_part_pos3( 0 ), folded.value() ).is_null() ) {
```

**Comment by Fris0uman** [comment]
reopening since it's stale because of feature freeze.

**Comment by kevingranade** [comment]
This still needs a clearer rationale for what is and is not foldable, I'm tempted to say that the footprint must be one tile, and that the foldable flag lets us take a vehicle that is multi-tile and treat it as being packed into one tile as per the original intent of the filding bicycle support.

**Comment by kevingranade** [comment]
As per comments this needs extensive rework, so closing for now, but do feel free to reopen once those are addressed.

**Pull request closed without merging**